tests/integration/hsm-integration-test.cpp

```diff
@@ -0,0 +1,452 @@
+                 const std::vector<uint8_t> &signature,
+                 const std::vector<uint8_t> &message)
+{
+    std::shared_ptr<MessageVerificationCtx> verifyCtx;
+    try {
+        verifyCtx = std::make_shared<ECDSASignaturePublicKeyCtx>(pubKey, digestType, sigFormat);
+    } catch (const MoCOCrWException &e) {
+        std::cerr << "Please check your ECC key. Failure creating context." << std::endl;
+        std::cerr << e.what();
+        exit(EXIT_FAILURE);
+    }
+    try {
+        verifyCtx->verifyMessage(signature, message);
+    } catch (const MoCOCrWException &e) {
+        /* Possible reasons:
+         * - error in openssl (sign, padding, ...)
+         * - Hash function's digest size doesn't match the message's digest size
+         * - Invalid signature format set
+         * - Signature can't be parsed
+         * - Invalid signature
+         */
+        std::cerr << "Verification failed!" << std::endl;
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+}
+
+int main(void)
+{
+    // Information for engine loading and key management.
+    std::string id("pkcs11");
+    std::string modulePath("/usr/lib/softhsm/libsofthsm2.so");
+    std::string tokenLabel("token-label");
+    // Don't hardcode the pin in your application, this is just for demonstration purposes
+    std::string pin("1234");
+    HsmEngine hsmEngine(id, modulePath, tokenLabel, pin);
+    std::vector<uint8_t> message = utility::fromHex("deadbeef");
+
+    /************** ECC key generation and ECDSA **************/
+    std::vector<uint8_t> keyIDECC{};
+    std::string keyLabelECC("ecc-key-label");
+    auto ecdsaDigestType = DigestTypes::SHA512;
+    ECCSpec ecspec;
+    auto eccPrivKey =
+            AsymmetricPrivateKey::generateKeyOnHSM(hsmEngine, ecspec, keyLabelECC, keyIDECC);
+    auto ecdsaSigFormat = ECDSASignatureFormat::ASN1_SEQUENCE_OF_INTS;
+
+    /* The argument hashFunction is optional. Default is SHA256
+     * The default signature format is ECDSASignatureFormat::ASN1_SEQUENCE_OF_INTS */
+    auto ECDSAsignature = ecdsaSign(eccPrivKey, ecdsaDigestType, ecdsaSigFormat, message);
+
+    /* Use-Case 1: You want to check your own signature.
+     * We can use here the private key, as it also contains the public key.
+     * In MoCOCrW the AsymmetricPrivateKey is a specialisation of AsymmetricPublicKey. Thus
+     * we do an implicit upcast here.
+     */
+    ecdsaVerify(eccPrivKey, ecdsaDigestType, ecdsaSigFormat, ECDSAsignature, message);
+
+    /* Use-Case 2: The public key used for verification is stored in the HSM */
+    auto pubKeyEcc = AsymmetricPublicKey::readPublicKeyFromHSM(hsmEngine, keyLabelECC, keyIDECC);
+    ecdsaVerify(pubKeyEcc, ecdsaDigestType, ecdsaSigFormat, ECDSAsignature, message);
+
+    /* Use-Case 3: You want to write the public key to a PEM file and use it later
+     * for verification
+     */
+    auto pubKeyPem = eccPrivKey.publicKeyToPem();
```

Comment:
This is now definitely a SW verification. So we can consider this as a test that checks if Crypto Inside/outside of the HSM plays nicely together.

Reply:
Agreed.